### ISSUE TITLE
`config_setting` false positive when `label_flag` points at an `alias`

### ISSUE THREAD

**Issue by tpudlik** [bazelbuild/bazel#21189]
### Description of the bug:

If a `config_setting` is set up to detect whether a `label_flag` points to a particular target, it will not fire if that target is pointed to via an alias.

### Which category does this issue belong to?

Configurability

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Consider the following BUILD file, and assume a file `main.cc` exists.

```
label_flag(
    name = "testflag",
    build_setting_default = ":one",
)

config_setting(
    name = "testflag_one",
    flag_values = {
        "//:testflag": "//:one",
    },
)

config_setting(
    name = "testflag_two",
    flag_values = {
        "//:testflag": "//:two",
    },
)

cc_library(
    name = "one",
)

cc_library(
    name = "two",
)

alias(
    name = "alias_for_one",
    actual = ":one",
)

cc_binary(
    name = "which",
    srcs = select({
        ":testflag_one": ["main.cc"],
        ":testflag_two": ["doesntexist.cc"],
        "//conditions:default": ["alsodoesntexist.cc"],
    }),
)
```

Running,

```
bazel build  --//:testflag=:one :which
```

builds `main.cc`, as expected. Running,

```
bazel build  --//:testflag=:two :which
```
returns an error, `missing input file '//:doesntexist.cc'`. This is also expected.

Running,

```
bazel build --/:testflag=:alias_for_one :which
```

returns an error, `missing input file '//:alsodoesntexist.cc'`. This is *unexpected*. The `testflag` does point to the expected target, it just does so through an alias.

### Which operating system are you running Bazel on?

Linux

### What is the output of `bazel info release`?

release 8.0.0-pre.20240108.6

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.

No, I don't believe this ever worked.

### Have you found anything relevant by searching the web?

This is related to https://github.com/bazelbuild/bazel/issues/13463. That issue was about false negatives arising when the _key_ in `config_setting.flag_value` was an alias. This issue is about false negatives when the _value_ in `config_setting.flag_values` is an alias.

### Any other information, logs, or outputs that you want to share?

I'd appreciate if someone knowledgeable (@lberki ?) could comment on whether this is a relatively trivial issue, or profoundly intractable.

**Comment by tpudlik** [comment]
I have a doc that discusses the context for this in some detail (internal,
though for no good reason:
https://docs.google.com/document/d/1O4xjnQBDpOxCMhlyzsowfYF3Cjq0fOfWB6hHsmsh-qI/edit?resourcekey=0-0B-fT2s05UYoC4TQIGDyvw&tab=t.0).
It's problematic because it makes it hard to integrate label flags with
other build configuration machinery like selects and
target_compatible_with. A workaround is to switch to constraint_settings as
soon as possible (option 2 in the doc), but that's error prone and involves
a lot of boilerplate.

**Comment by fmeum** [comment]
@tpudlik Could you share a public version (or a copy) of the doc?

**Comment by tpudlik** [comment]
The main practical motivation for such `config_setting`s based on `label_flag`s is to do the following:

1. Mark targets (especially tests) only `target_compatible_with` given possible targets that the `label_flag` might point to.

    We could work around this by directly depending on the relevant `label_flag` value from the test. (This breaks down if there's more than one value that you're compatible with, but in practice maybe that's good enough.)

2. Use a different source file for a library based on which backend is being used (probably using a `select`, here's [an example](https://cs.opensource.google/pigweed/pigweed/+/main:pw_unit_test/BUILD.bazel;l=290-318;drc=6b853a8137d3b9d771c35dae26fd56a208cb921c) of this being done).

    Maybe we should just reorganize the BUILD file (break up some targets) so that the `select`s in (2) are not needed. Hmm let me try.

But beyond these specific use cases, it would generally be nice to be able to configure your build based on `label_flag` choices (in the same way that you can configure it based on `constraint_value` choices in a platform).

**Comment by gregestren** [comment]
I don't think this is unsolvable. I agree with Lukacs' concerns that there are consequences, so solving has costs.

You could argue that rule writers are responsible for those costs: don't set up `config_setting` / `label_flag` / `alias` scenarios that trigger expensive loads.

The implementation costs are extra Blaze complexity.  I think that's solvable at a moderate complexity cost (like pre-check `config_setting` rules before evaluating and Skyframe-evaluate label-flag values if encountered, then pass that to the `config_setting` rule implementation). That's a non-trivial amount of complexity.

All this makes me want to see if we can find a user space solution instead.

**Comment by gregestren** [comment]
I suggested in the doc that [user-defined config settings](https://docs.google.com/document/d/1p02Y9joQSgdXtiTA2mJ_UXHBiBDPPTqrO-oJcupLAu8/edit#heading=h.icwoewbtra8) might resolve this, and that's something we already want to do.

It doesn't address the eager dependency resolution concern. But that's intrinsic to the problem space.
